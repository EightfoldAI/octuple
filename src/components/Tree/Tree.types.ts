import type {
  BasicDataNode,
  DataNode,
  Key,
  OcTreeProps,
} from './Internal/OcTree.types';

export type SwitcherIcon =
  | React.ReactNode
  | ((props: { expanded: boolean }) => React.ReactNode);

export interface OcTreeNodeAttribute {
  eventKey: string;
  classNames: string;
  expanded: boolean;
  selected: boolean;
  checked: boolean;
  halfChecked: boolean;
  children: React.ReactNode;
  title: React.ReactNode;
  pos: string;
  dragOver: boolean;
  dragOverGapTop: boolean;
  dragOverGapBottom: boolean;
  isLeaf: boolean;
  selectable: boolean;
  disabled: boolean;
  disableCheckbox: boolean;
}

export interface OcTreeNodeProps {
  classNames?: string;
  checkable?: boolean;
  disabled?: boolean;
  disableCheckbox?: boolean;
  title?: string | React.ReactNode;
  key?: Key;
  eventKey?: string;
  isLeaf?: boolean;
  checked?: boolean;
  expanded?: boolean;
  loading?: boolean;
  selected?: boolean;
  selectable?: boolean;
  icon?: ((treeNode: OcTreeNodeAttribute) => React.ReactNode) | React.ReactNode;
  children?: React.ReactNode;
  [customProp: string]: any;
}

export interface OcTreeNode extends React.Component<OcTreeNodeProps, {}> {}

export interface OcTreeNodeBaseEvent {
  node: OcTreeNode;
  nativeEvent: MouseEvent;
}

export interface OcTreeNodeCheckedEvent extends OcTreeNodeBaseEvent {
  event: 'check';
  checked?: boolean;
  checkedNodes?: OcTreeNode[];
}

export interface OcTreeNodeSelectedEvent extends OcTreeNodeBaseEvent {
  event: 'select';
  selected?: boolean;
  selectedNodes?: DataNode[];
}

export interface OcTreeNodeExpandedEvent extends OcTreeNodeBaseEvent {
  expanded?: boolean;
}

export interface OcTreeNodeMouseEvent {
  node: OcTreeNode;
  event: React.DragEvent<HTMLElement>;
}

export interface OcTreeNodeDragEnterEvent extends OcTreeNodeMouseEvent {
  expandedKeys: Key[];
}

export interface OcTreeNodeDropEvent {
  node: OcTreeNode;
  dragNode: OcTreeNode;
  dragNodesKeys: Key[];
  dropPosition: number;
  dropToGap?: boolean;
  event: React.MouseEvent<HTMLElement>;
}

// [Legacy] Compatible for v3
export type TreeNodeNormal = DataNode;

export type DraggableFn = (node: OcTreeNode) => boolean;

export interface DraggableConfig {
  icon?: React.ReactNode | false;
  nodeDraggable?: DraggableFn;
}

export interface TreeProps<T extends BasicDataNode = DataNode>
  extends Omit<OcTreeProps<T>, 'showLine' | 'direction' | 'draggable'> {
  showLine?: boolean | { showLeafIcon: boolean };
  classNames?: string;
  multiple?: boolean;
  autoExpandParent?: boolean;
  checkStrictly?: boolean;
  checkable?: boolean;
  disabled?: boolean;
  defaultExpandAll?: boolean;
  defaultExpandParent?: boolean;
  defaultExpandedKeys?: Key[];
  expandedKeys?: Key[];
  checkedKeys?: Key[] | { checked: Key[]; halfChecked: Key[] };
  defaultCheckedKeys?: Key[];
  selectedKeys?: Key[];
  defaultSelectedKeys?: Key[];
  selectable?: boolean;
  filterOcTreeNode?: (node: OcTreeNode) => boolean;
  loadedKeys?: Key[];
  draggable?: DraggableFn | boolean | DraggableConfig;
  style?: React.CSSProperties;
  showIcon?: boolean;
  icon?:
    | ((nodeProps: OcTreeNodeAttribute) => React.ReactNode)
    | React.ReactNode;
  switcherIcon?: SwitcherIcon;
  children?: React.ReactNode;
  blockNode?: boolean;
  virtual?: boolean;
}
