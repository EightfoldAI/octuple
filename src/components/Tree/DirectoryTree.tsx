import React from 'react';
import { debounce, mergeClasses } from '../../shared/utilities';
import type OcTree from './Internal';
import { conductExpandParent } from './Internal/util';
import type { EventDataNode, DataNode, Key } from './Internal/OcTree.types';
import {
  convertDataToEntities,
  convertTreeToData,
} from './Internal/utils/treeUtil';
import type { TreeProps, OcTreeNodeAttribute } from './Tree.types';
import Tree from './Tree';
import { calcRangeKeys, convertDirectoryKeysToNodes } from './Utils/dictUtil';
import { Icon, IconName, IconSize } from '../Icon';
import { useCanvasDirection } from '../../hooks/useCanvasDirection';

import styles from './Styles/tree.module.scss';

export type ExpandAction = false | 'click' | 'doubleClick';

export interface DirectoryTreeProps extends TreeProps {
  expandAction?: ExpandAction;
}

export interface DirectoryTreeState {
  expandedKeys?: Key[];
  selectedKeys?: Key[];
}

function getIcon(props: OcTreeNodeAttribute): React.ReactNode {
  const { isLeaf, expanded } = props;
  if (isLeaf) {
    return (
      <Icon path={IconName.mdiFileDocumentOutline} size={IconSize.Small} />
    );
  }
  return expanded ? (
    <Icon path={IconName.mdiFolderOpenOutline} size={IconSize.Small} />
  ) : (
    <Icon path={IconName.mdiFolderOutline} size={IconSize.Small} />
  );
}

function getTreeData({ treeData, children }: DirectoryTreeProps) {
  return treeData || convertTreeToData(children);
}

const DirectoryTree: React.ForwardRefRenderFunction<
  OcTree,
  DirectoryTreeProps
> = (
  { defaultExpandAll, defaultExpandParent, defaultExpandedKeys, ...props },
  ref
) => {
  // Shift click usage
  const lastSelectedKey = React.useRef<Key>();

  const cachedSelectedKeys = React.useRef<Key[]>();

  const treeRef = React.createRef<OcTree>();

  React.useImperativeHandle(ref, () => treeRef.current!);

  const getInitExpandedKeys = () => {
    const { keyEntities } = convertDataToEntities(getTreeData(props));

    let initExpandedKeys: any;

    // Expanded keys
    if (defaultExpandAll) {
      initExpandedKeys = Object.keys(keyEntities);
    } else if (defaultExpandParent) {
      initExpandedKeys = conductExpandParent(
        props.expandedKeys || defaultExpandedKeys || [],
        keyEntities
      );
    } else {
      initExpandedKeys = props.expandedKeys || defaultExpandedKeys;
    }
    return initExpandedKeys;
  };

  const [selectedKeys, setSelectedKeys] = React.useState(
    props.selectedKeys || props.defaultSelectedKeys || []
  );
  const [expandedKeys, setExpandedKeys] = React.useState(getInitExpandedKeys());

  React.useEffect(() => {
    if ('selectedKeys' in props) {
      setSelectedKeys(props.selectedKeys!);
    }
  }, [props.selectedKeys]);

  React.useEffect(() => {
    if ('expandedKeys' in props) {
      setExpandedKeys(props.expandedKeys);
    }
  }, [props.expandedKeys]);

  const expandFolderNode = (
    event: React.MouseEvent<HTMLElement>,
    node: any
  ) => {
    const { isLeaf } = node;

    if (isLeaf || event.shiftKey || event.metaKey || event.ctrlKey) {
      return;
    }

    // Call internal tree expand function
    debounce(() => {
      treeRef.current!.onNodeExpand(event as any, node);
    }, 200);
  };

  const onExpand = (
    keys: Key[],
    info: {
      node: EventDataNode<DataNode>;
      expanded: boolean;
      nativeEvent: MouseEvent;
    }
  ) => {
    if (!('expandedKeys' in props)) {
      setExpandedKeys(keys);
    }
    // Call origin function
    return props.onExpand?.(keys, info);
  };

  const onClick = (
    event: React.MouseEvent<HTMLElement>,
    node: EventDataNode<DataNode>
  ) => {
    const { expandAction } = props;

    // Expand the tree
    if (expandAction === 'click') {
      expandFolderNode(event, node);
    }

    props.onClick?.(event, node);
  };

  const onDoubleClick = (
    event: React.MouseEvent<HTMLElement>,
    node: EventDataNode<DataNode>
  ) => {
    const { expandAction } = props;

    // Expand the tree
    if (expandAction === 'doubleClick') {
      expandFolderNode(event, node);
    }

    props.onDoubleClick?.(event, node);
  };

  const onSelect = (
    keys: Key[],
    event: {
      event: 'select';
      selected: boolean;
      node: any;
      selectedNodes: DataNode[];
      nativeEvent: MouseEvent;
    }
  ) => {
    const { multiple } = props;
    const { node, nativeEvent } = event;
    const { key = '' } = node;

    const treeData = getTreeData(props);
    // const newState: DirectoryTreeState = {};

    // We need wrap this event since some value is not same
    const newEvent: any = {
      ...event,
      selected: true, // Directory selected always true
    };

    // Windows / Mac single pick
    const ctrlPick: boolean = nativeEvent?.ctrlKey || nativeEvent?.metaKey;
    const shiftPick: boolean = nativeEvent?.shiftKey;

    // Generate new selected keys
    let newSelectedKeys: Key[];
    if (multiple && ctrlPick) {
      // Control click
      newSelectedKeys = keys;
      lastSelectedKey.current = key;
      cachedSelectedKeys.current = newSelectedKeys;
      newEvent.selectedNodes = convertDirectoryKeysToNodes(
        treeData,
        newSelectedKeys
      );
    } else if (multiple && shiftPick) {
      // Shift click
      newSelectedKeys = Array.from(
        new Set([
          ...(cachedSelectedKeys.current || []),
          ...calcRangeKeys({
            treeData,
            expandedKeys,
            startKey: key,
            endKey: lastSelectedKey.current,
          }),
        ])
      );
      newEvent.selectedNodes = convertDirectoryKeysToNodes(
        treeData,
        newSelectedKeys
      );
    } else {
      // Single click
      newSelectedKeys = [key];
      lastSelectedKey.current = key;
      cachedSelectedKeys.current = newSelectedKeys;
      newEvent.selectedNodes = convertDirectoryKeysToNodes(
        treeData,
        newSelectedKeys
      );
    }

    props.onSelect?.(newSelectedKeys, newEvent);
    if (!('selectedKeys' in props)) {
      setSelectedKeys(newSelectedKeys);
    }
  };
  const htmlDir: string = useCanvasDirection();

  const { classNames, ...otherProps } = props;

  const connectClassNames = mergeClasses([
    styles.treeDirectory,
    { ['tree-directory-rtl']: htmlDir === 'rtl' },
    classNames,
  ]);

  return (
    <Tree
      icon={getIcon}
      ref={treeRef}
      blockNode
      {...otherProps}
      classNames={connectClassNames}
      expandedKeys={expandedKeys}
      selectedKeys={selectedKeys}
      onSelect={onSelect}
      onClick={onClick}
      onDoubleClick={onDoubleClick}
      onExpand={onExpand}
    />
  );
};

const ForwardDirectoryTree = React.forwardRef(DirectoryTree);
ForwardDirectoryTree.displayName = 'DirectoryTree';

ForwardDirectoryTree.defaultProps = {
  showIcon: true,
  expandAction: 'click' as DirectoryTreeProps['expandAction'],
};

export default ForwardDirectoryTree;
