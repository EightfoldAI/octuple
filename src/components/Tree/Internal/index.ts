import Tree from './OcTree';
import TreeNode from './TreeNode';
import type {
  BasicDataNode,
  FieldDataNode,
  TreeNodeProps,
  OcTreeProps,
} from './OcTree.types';

export { TreeNode };
export type { OcTreeProps, TreeNodeProps, BasicDataNode, FieldDataNode };
export default Tree;
