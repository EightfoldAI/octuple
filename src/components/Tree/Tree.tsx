'use client';

import React from 'react';
import OcTree, { TreeNode } from './Internal';
import { DraggableConfig, OcTreeNodeProps, TreeProps } from './Tree.types';
import type { BasicDataNode, DataNode } from './Internal/OcTree.types';
import DirectoryTree from './DirectoryTree';
import renderSwitcherIcon from './Utils/iconUtil';
import dropIndicatorRender from './Utils/dropIndicator';
import { Icon, IconName, IconSize } from '../Icon';
import { useCanvasDirection } from '../../hooks/useCanvasDirection';
import { collapseMotion, mergeClasses } from '../../shared/utilities';

import styles from './Styles/tree.module.scss';

type CompoundedComponent = (<T extends BasicDataNode | DataNode = DataNode>(
  props: React.PropsWithChildren<TreeProps<T>> & { ref?: React.Ref<OcTree> }
) => React.ReactElement) & {
  defaultProps: Partial<React.PropsWithChildren<TreeProps<any>>>;
  TreeNode: typeof TreeNode;
  DirectoryTree: typeof DirectoryTree;
};

const Tree = React.forwardRef<OcTree, TreeProps>((props, ref) => {
  const {
    classNames,
    showIcon,
    showLine,
    switcherIcon,
    blockNode,
    children,
    checkable,
    selectable,
    draggable,
    virtual,
  } = props;
  const htmlDir: string = useCanvasDirection();
  const newProps = {
    ...props,
    showLine: Boolean(showLine),
    dropIndicatorRender,
  };

  const draggableConfig = React.useMemo(() => {
    if (!draggable) {
      return false;
    }

    let mergedDraggable: DraggableConfig = {};
    switch (typeof draggable) {
      case 'function':
        mergedDraggable.nodeDraggable = draggable;
        break;

      case 'object':
        mergedDraggable = { ...draggable };
        break;

      default:
      // Do nothing
    }

    if (mergedDraggable.icon !== false) {
      mergedDraggable.icon = mergedDraggable.icon || (
        <Icon
          color={'#ffffff'}
          path={IconName.mdiHandBackRight}
          size={IconSize.Medium}
        />
      );
    }

    return mergedDraggable;
  }, [draggable]);

  return (
    <OcTree
      itemHeight={20}
      ref={ref}
      virtual={virtual}
      {...newProps}
      classNames={mergeClasses([
        { ['tree-icon-hide']: !showIcon },
        { [styles.treeBlockNode]: blockNode },
        { [styles.treeUnselectable]: !selectable },
        { [styles.treeRtl]: htmlDir === 'rtl' },
        classNames,
      ])}
      direction={htmlDir}
      checkable={
        checkable ? <span className={'tree-checkbox-inner'} /> : checkable
      }
      selectable={selectable}
      switcherIcon={(nodeProps: OcTreeNodeProps) =>
        renderSwitcherIcon(switcherIcon, showLine, nodeProps)
      }
      draggable={draggableConfig as any}
    >
      {children}
    </OcTree>
  );
}) as unknown as CompoundedComponent;

Tree.TreeNode = TreeNode;

Tree.DirectoryTree = DirectoryTree;

Tree.defaultProps = {
  checkable: false,
  selectable: true,
  showIcon: false,
  motion: {
    ...collapseMotion,
    motionAppear: false,
  },
  blockNode: false,
};

export default Tree;
