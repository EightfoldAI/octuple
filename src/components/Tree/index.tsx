import Tree from './Tree';

export { EventDataNode, DataNode } from './Internal/OcTree.types';

export {
    TreeProps,
    OcTreeNode,
    OcTreeNodeMouseEvent,
    OcTreeNodeExpandedEvent,
    OcTreeNodeCheckedEvent,
    OcTreeNodeSelectedEvent,
    OcTreeNodeAttribute,
    OcTreeNodeProps,
} from './Tree.types';

export {
    ExpandAction as DirectoryTreeExpandAction,
    DirectoryTreeProps,
} from './DirectoryTree';

export default Tree;
