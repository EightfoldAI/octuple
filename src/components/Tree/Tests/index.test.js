import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import Tree from '../index';
import 'jest-specific-snapshot';

Enzyme.configure({ adapter: new Adapter() });

const { TreeNode } = Tree;

describe('Tree', () => {
  it('icon and switcherIcon of Tree with showLine should render correctly', () => {
    const wrapper = mount(
      <Tree showLine showIcon>
        <TreeNode icon="icon" switcherIcon="switcherIcon" key="0-0">
          <TreeNode icon="icon" switcherIcon="switcherIcon" key="0-0-0" />
          <TreeNode switcherIcon="switcherIcon" key="0-0-1" />
          <TreeNode icon="icon" key="0-0-2" />
          <TreeNode key="0-0-3" />
        </TreeNode>
        <TreeNode switcherIcon="switcherIcon" key="0-1">
          <TreeNode icon="icon" switcherIcon="switcherIcon" key="0-0-0" />
          <TreeNode switcherIcon="switcherIcon" key="0-0-1" />
          <TreeNode icon="icon" key="0-0-2" />
          <TreeNode key="0-0-3" />
        </TreeNode>
        <TreeNode key="0-2">
          <TreeNode icon="icon" switcherIcon="switcherIcon" key="0-0-0" />
          <TreeNode switcherIcon="switcherIcon" key="0-0-1" />
          <TreeNode icon="icon" key="0-0-2" />
          <TreeNode key="0-0-3" />
        </TreeNode>
      </Tree>
    );
    expect(wrapper.render()).toMatchSpecificSnapshot(
      './__snapshots__/index.treeiconrender.shot'
    );
  });

  it('switcherIcon in Tree should not render at leaf nodes', () => {
    const wrapper = mount(
      <Tree switcherIcon={<i className="switcherIcon" />} defaultExpandAll>
        <TreeNode icon="icon">
          <TreeNode id="node1" title="node1" icon="icon" key="0-0-2" />
          <TreeNode id="node2" title="node2" key="0-0-3" />
        </TreeNode>
      </Tree>
    );
    expect(wrapper.find('.switcherIcon').length).toBe(1);
  });

  it('switcherIcon in Tree could be string', () => {
    const wrapper = mount(
      <Tree switcherIcon="switcherIcon" defaultExpandAll>
        <TreeNode icon="icon">
          <TreeNode id="node1" title="node1" icon="icon" key="0-0-2" />
          <TreeNode id="node2" title="node2" key="0-0-3" />
        </TreeNode>
      </Tree>
    );
    expect(wrapper.render()).toMatchSpecificSnapshot(
      './__snapshots__/index.switchericonstring.shot'
    );
  });

  it('switcherIcon should be loading icon when loadData', () => {
    const onLoadData = () =>
      new Promise((resolve) => {
        setTimeout(() => {
          resolve();
        }, 1000);
      });
    const wrapper = mount(
      <Tree switcherIcon="switcherIcon" defaultExpandAll loadData={onLoadData}>
        <TreeNode icon="icon" id="node1" key="0-0" title="node1" />
      </Tree>
    );
    expect(wrapper.render()).toMatchSpecificSnapshot(
      './__snapshots__/index.loadingicon.shot'
    );
  });

  it('switcherIcon in Tree could be render prop function', () => {
    const wrapper = mount(
      <Tree
        switcherIcon={(expanded) =>
          expanded ? <span className="open" /> : <span className="close" />
        }
        defaultExpandAll
      >
        <TreeNode icon="icon">
          <TreeNode id="node1" title="node1" icon="icon" key="0-0-2" />
          <TreeNode id="node2" title="node2" key="0-0-3" />
        </TreeNode>
      </Tree>
    );
    expect(wrapper.find('.open').length).toBe(1);
  });

  it('showLine is object type should render correctly', () => {
    const wrapper = mount(
      <Tree showLine={{ showLeafIcon: false }} defaultExpandedKeys={['0-0-0']}>
        <TreeNode title="parent 1" key="0-0">
          <TreeNode title="parent 1-0" key="0-0-0">
            <TreeNode title="leaf" key="0-0-0-0" />
          </TreeNode>
        </TreeNode>
      </Tree>
    );
    expect(wrapper.render()).toMatchSpecificSnapshot(
      './__snapshots__/index.showline.shot'
    );
  });

  describe('draggable', () => {
    const dragTreeData = [
      {
        title: 'bamboo',
        key: 'bamboo',
      },
    ];

    it('hide icon', () => {
      const wrapper = mount(
        <Tree treeData={dragTreeData} draggable={{ icon: false }} />
      );
      expect(wrapper.exists('.anticon-holder')).toBeFalsy();
    });

    it('customize icon', () => {
      const wrapper = mount(
        <Tree
          treeData={dragTreeData}
          draggable={{ icon: <span className="little" /> }}
        />
      );
      expect(wrapper.exists('.little')).toBeTruthy();
    });

    it('nodeDraggable', () => {
      const nodeDraggable = jest.fn(() => false);
      mount(<Tree treeData={dragTreeData} draggable={{ nodeDraggable }} />);
      expect(nodeDraggable).toHaveBeenCalledWith(dragTreeData[0]);
    });

    it('nodeDraggable func', () => {
      const nodeDraggable = jest.fn(() => false);
      mount(<Tree treeData={dragTreeData} draggable={nodeDraggable} />);
      expect(nodeDraggable).toHaveBeenCalledWith(dragTreeData[0]);
    });
  });
});
