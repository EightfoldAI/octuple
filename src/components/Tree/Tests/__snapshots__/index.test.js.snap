version https://git-lfs.github.com/spec/v1
oid sha256:d9582a6ea7b772ed3f54881ccdee762491d2b8e38cbbc240eaf73a41c46fef37
size 7835601
